R/hub-connection.R

```diff
@@ -0,0 +1,178 @@
+#'   extracting or submitting data.
+#' @export
+#'
+#' @examples
+#' # Simple forecasting Hub example
+#' hub_path <- system.file("hub_1", package = "hubUtils")
+#' connect_hub(hub_path)
+#' # Scenario Hub example
+#' scnr_path <- system.file("scnr_hub_1", package = "hubUtils")
+#' connect_hub(hub_path)
+connect_hub <- function(hub_path, hubmeta_path = NULL,
+                        hubmeta_format = c("json", "yaml", "yml")) {
+
+    hubmeta_format <- rlang::arg_match(hubmeta_format)
+
+    if (is.null(hubmeta_path)) {
+        hubmeta_path <- fs::path(
+            hub_path,
+            "hubmeta",
+            ext = hubmeta_format
+            )
+    }
+
+    hubmeta <- read_hubmeta(hubmeta_path)
+
+    hubmeta[["hubmeta_path"]] <- hubmeta_path
+    hubmeta[["hub_path"]] <- hub_path
+
+    new_hub_connection(hubmeta)
+
+}
+
+# INTERNAL ----
+
+#' Create new hub connection S3 class object
+#'
+#' Internal function for create a new hub connection S3 class object
+#' @param x A metadata list generated by [read_hubmeta()].
+#'
+#' @return A hub connection S3 class object.
+#' @noRd
+new_hub_connection <- function(x) {
+    stopifnot(is.list(x))
+
+    x <- assing_hc_attrs(x)
+
+    class(x) <- "hub_connection"
+
+    x
+}
+
+
+#' Assign high level hub metadata as attributes.
+#'
+#' @param x A metadata list generated by [read_hubmeta()].
+#'
+#' @return
+#'   A metadata list with `task_ids` names and `round_ids` assigned to
+#'   attributes `task_id_names` and `round_ids` respectively.
+#'
+#'   Where the metadata list contains round specific metadata
+#'   elements, a named list containing `task_ids` names per round is assigned to
+#'   `task_id_names` and a character vector of the names of elements containing
+#'   round specific metadata is assigned to `round_ids`.
+#'
+#'   Where metadata are shared between rounds
+#'   (indicated by the name of the first element of the metadata list being
+#'   `"round_id_from_variable"`), a single character vector of task_ids names
+#'   is assigned to attribute `task_id_names` and the value of element
+#'   `"round_id_from_variable"` is assigned to attribute `round_ids`.
+#'
+#'   Where a `hubmeta_path` and/or `hub_path` elements are present in the metadata
+#'   list, they are removed from the list and assigned to attributes `hubmeta_path`
+#'   `hub_path` respectively.
+#' @noRd
+assing_hc_attrs <- function(x) {
+
+    # if assigned, move hubmeta_path & hub_path to attributes to maintain
+    # original hubmeta structure
+    if(!is.null(x$hubmeta_path)){
+        attr(x, "hubmeta_path") <- x$hubmeta_path
+        x$hubmeta_path <- NULL
+    }
+
+    if(!is.null(x$hub_path)){
+        attr(x, "hub_path") <- x$hub_path
+        x$hub_path <- NULL
+    }
+
+    if (names(x)[1] == "round_id_from_variable") {
+
+        task_id_names <- purrr::pluck(
+            x,
+            "round_id_from_variable",
+            "model_tasks",
+            1,
```

Comment:
Same comment as elsewhere that I think we don't want to pull only entries at index 1

Reply:
See comment above